File: TinyScript/JS/src/translator/symbol/Symbol.js
const SymbolType = require('./SymbolType')

class Symbol {
  // 本来这是应该要做3个类的，但是这样太麻烦了，所以就合并成一个类了
  // 这种就是 union
  // 类型少时才推荐使用，不然字段会变得很多，很臃肿
  constructor(type) {
    this.type = type
    this.label = null
    this.offset = 0
    this.layerOffset = 0 // 用于记录层级作用域
    this.lexeme = null
    this.parent = null
  }

  // 创建三种 symbol 工厂方法
  // 加了 static 声明的作用是：在别的文件里，可通过通过 Symbol.函数名 直接使用
  // 因为 static 就是指在编译后所分配的内存会一直存在，直到程序退出内存才会释放这个空间，也就是只要程序在运行，那么这块内存就会一直存在。
  static createAddressSymbol(lexeme, offset) {
    var symbol = new Symbol(SymbolType.ADDRESS_SYMBOL)
    symbol.lexeme = lexeme
    symbol.offset = offset
    return symbol
  }

  static createImmediateSymbol(lexeme) {
    var symbol = new Symbol(SymbolType.IMMEDIATE_SYMBOL)
    symbol.lexeme = lexeme
    return symbol
  }

  static createLabelSymbol(label, lexeme) {
    var symbol = new Symbol(SymbolType.ADDRESS_SYMBOL)
    symbol.label = label
    symbol.lexeme = lexeme
    return symbol
  }

  // 拷贝一份 symbol 函数
  copy() {
    var symbol = new Symbol(this.type)
    symbol.lexeme = this.lexeme
    symbol.label = this.label
    symbol.offset = this.offset
    symbol.layerOffset = this.layerOffset
    symbol.type = this.type
    return symbol
  }

  // 下面就是一些 设置/获取 基础字段属性函数
  setParent(parent) {
    this.parent = parent
  }

  setOffset(offset) {
    this.offset = offset
  }

  getOffset() {
    return this.offset
  }

  getType() {
    return this.type
  }

  toString() {
    if (this.type === SymbolType.LABEL_SYMBOL) {
      return this.label;
    }
    return this.lexeme.getValue()
  }

  setLexeme(lexeme) {
    this.lexeme = lexeme
  }

  getLexeme() {
    return this.lexeme
  }

  setLayerOffset(offset) {
    this.layerOffset = offset
  }

  getLayerOffset() {
    return this.layerOffset
  }

  getLabel() {
    return this.label
  }
}

module.exports = Symbol

File: TinyScript/JAVA/src/translator/symbol/Symbol.java
package translator.symbol;

import lexer.Token;

public class Symbol {
    // union 联合设置
    SymbolTable parent;
    Token lexeme;
    String label;
    int offset;
    int layerOffset = 0; // 用来记录层级关系的 offset
    SymbolType type;

    public Symbol(SymbolType type) {
        this.type = type;
    }

    // 三个工厂构造函数
    public static Symbol createAddressSymbol(Token lexeme, int offset) {
        var symbol = new Symbol(SymbolType.ADDRESS_SYMBOL);
        symbol.lexeme = lexeme;
        symbol.offset = offset;
        return symbol;
    }

    public static Symbol createImmediateSymbol(Token lexeme) {
        var symbol = new Symbol(SymbolType.IMMEDIATE_SYMBOL);
        symbol.lexeme = lexeme;
        return symbol;
    }

    public static Symbol createLabelSymbol(String label, Token lexeme) {
        var symbol = new Symbol(SymbolType.LABEL_SYMBOL);
        symbol.lexeme = lexeme;
        symbol.label = label;
        return symbol;
    }

    // 把（上面构造函数的）符号彻底的复制一份
    public Symbol copy() {
        var symbol = new Symbol(this.type);
        symbol.lexeme = this.lexeme;
        symbol.label = this.label;
        symbol.offset = this.offset;
        symbol.layerOffset = this.layerOffset;
        symbol.type = this.type;
        return symbol;
    }

    // 一些常规的方法
    // 设置 parent
    public void serParent(SymbolTable parent) {
        this.parent = parent;
    }

    public void setOffset(int offset) {
        this.offset = offset;
    }

    public int getOffset() {
        return this.offset;
    }

    public SymbolType getType() {
        return this.type;
    }

    @Override
    public String toString() {
        switch (this.type) {
            case ADDRESS_SYMBOL:
            case LABEL_SYMBOL:
                return lexeme.getValue();
            case IMMEDIATE_SYMBOL:
                return label;
        }
        return "";
    }

    public void setLexeme(Token lexeme) {
        this.lexeme = lexeme;
    }


    public Token getLexeme() {
        return this.lexeme;
    }

    public void setLayerOffset(int offset) {
        this.layerOffset = offset;
    }

    public int getLayerOffset() {
        return this.layerOffset;
    }

    public String getLabel() {
        return this.label;
    }
}
